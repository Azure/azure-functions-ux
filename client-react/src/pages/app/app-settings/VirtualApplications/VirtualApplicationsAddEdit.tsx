import * as React from 'react';
import { TextField } from 'office-ui-fabric-react/lib-commonjs/TextField';
import { VirtualApplication } from '../../../../models/WebAppModels';
import { Toggle } from 'office-ui-fabric-react/lib-commonjs/Toggle';

export interface HandlerMappingAddEditProps extends VirtualApplication {
  updateVirtualApplication: (item: VirtualApplication) => any;
}

const VirtualApplicationsAddEdit: React.SFC<HandlerMappingAddEditProps> = props => {
  const { updateVirtualApplication, children, ...virtualApplication } = props;
  const updatePhysicalPath = (physicalPath: string) => {
    updateVirtualApplication({
      ...virtualApplication,
      physicalPath,
    });
  };

  const updateVirtualPath = (virtualPath: string) => {
    updateVirtualApplication({ ...virtualApplication, virtualPath });
  };

  const updateVirtualDirectory = (virtualDirectory: boolean) => {
    updateVirtualApplication({
      ...virtualApplication,
      virtualDirectory,
    });
  };
  const updatePreloadEnabled = (preloadEnabled: boolean) => {
    updateVirtualApplication({
      ...virtualApplication,
      preloadEnabled,
    });
  };
  return (
    <div>
      <TextField label="Physical Path" id="extension" value={virtualApplication.physicalPath} onChanged={updatePhysicalPath} />
      <TextField label="Virtual Path" id="value" value={virtualApplication.virtualPath} onChanged={updateVirtualPath} />
      <Toggle
        label="Directory or Application"
        id="sticky"
        defaultChecked={virtualApplication.virtualDirectory}
        onChanged={updateVirtualDirectory}
        onText="Directory"
        offText="Application"
      />
      {virtualApplication.virtualDirectory ? null : (
        <Toggle
          label="Preload Enabled"
          id="preloadEnabled"
          defaultChecked={virtualApplication.preloadEnabled}
          onChanged={updatePreloadEnabled}
          onText="On"
          offText="Off"
        />
      )}
    </div>
  );
};

export default VirtualApplicationsAddEdit;
