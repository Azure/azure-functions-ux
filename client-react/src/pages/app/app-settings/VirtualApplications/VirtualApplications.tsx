import * as React from 'react';
import { DetailsList, DetailsListLayoutMode, IColumn, SelectionMode } from 'office-ui-fabric-react/lib-commonjs/DetailsList';
import { IconButton, PrimaryButton, ActionButton, DefaultButton } from 'office-ui-fabric-react/lib-commonjs/Button';
import { AppSettingsFormValues } from '../AppSettings.Types';
import { Panel, PanelType } from 'office-ui-fabric-react/lib-commonjs/Panel';
import { VirtualApplication } from '../../../../models/WebAppModels';
import VirtualApplicationsAddEdit from './VirtualApplicationsAddEdit';
import { FormikProps } from 'formik';

export interface VirtualApplicationsState {
  showPanel: boolean;
  currentVirtualApplication: VirtualApplication | null;
  currentItemIndex: number | null;
  createNewItem: boolean;
}
export default class VirtualApplications extends React.Component<FormikProps<AppSettingsFormValues>, VirtualApplicationsState> {
  constructor(props) {
    super(props);
    this.state = {
      showPanel: false,
      currentVirtualApplication: null,
      currentItemIndex: null,
      createNewItem: false,
    };
  }

  public render() {
    const { values } = this.props;
    if (!values.virtualApplications) {
      return null;
    }
    return (
      <>
        <ActionButton onClick={this.createNewItem} styles={{ root: { marginTop: '5px' } }} iconProps={{ iconName: 'Add' }}>
          New Directory/Application
        </ActionButton>
        <Panel
          isOpen={this.state.showPanel}
          type={PanelType.medium}
          onDismiss={this._onClosePanel}
          headerText="New App "
          closeButtonAriaLabel="Close"
          onRenderFooterContent={this._onRenderFooterContent}>
          <VirtualApplicationsAddEdit {...this.state.currentVirtualApplication!} updateVirtualApplication={this.updateCurrentItem} />
        </Panel>
        <DetailsList
          items={values.virtualApplications}
          columns={this._columns}
          isHeaderVisible={true}
          layoutMode={DetailsListLayoutMode.justified}
          selectionMode={SelectionMode.none}
          selectionPreservedOnEmptyClick={true}
          ariaLabelForSelectionColumn="Toggle selection"
          ariaLabelForSelectAllCheckbox="Toggle selection for all items"
        />
      </>
    );
  }

  private updateCurrentItem = (item: VirtualApplication) => {
    this.setState({ currentVirtualApplication: item });
  };

  private createNewItem = () => {
    const blank: VirtualApplication = {
      physicalPath: '',
      virtualPath: '',
      virtualDirectories: [],
      preloadEnabled: false,
      virtualDirectory: true,
    };
    this.setState({
      showPanel: true,
      currentVirtualApplication: blank,
      createNewItem: true,
      currentItemIndex: -1,
    });
  };

  private _onClosePanel = (): void => {
    const { values, setValues } = this.props;
    const virtualApplications = [...values.virtualApplications];
    if (!this.state.createNewItem) {
      virtualApplications[this.state.currentItemIndex!] = this.state.currentVirtualApplication!;
    } else {
      virtualApplications.push(this.state.currentVirtualApplication!);
    }
    setValues({
      ...values,
      virtualApplications,
    });
    this.setState({ createNewItem: false, showPanel: false });
  };

  private _onRenderFooterContent = (): JSX.Element => {
    return (
      <div>
        <PrimaryButton onClick={this._onClosePanel} style={{ marginRight: '8px' }}>
          Save
        </PrimaryButton>
        <DefaultButton onClick={this._onClosePanel}>Cancel</DefaultButton>
      </div>
    );
  };
  private _onShowPanel = (item: VirtualApplication, index: number): void => {
    this.setState({
      showPanel: true,
      currentVirtualApplication: item,
      currentItemIndex: index,
    });
  };

  private removeItem(index: number) {
    const { values, setValues } = this.props;
    const virtualApplications: VirtualApplication[] = [...values.virtualApplications];
    virtualApplications.splice(index, 1);
    setValues({
      ...values,
      virtualApplications,
    });
  }

  private onRenderItemColumn = (item: VirtualApplication, index: number, column: IColumn) => {
    if (!column || !item) {
      return null;
    }

    if (column.key === 'delete') {
      return item.virtualPath === '/' ? null : (
        <IconButton iconProps={{ iconName: 'Delete' }} title="Delete" onClick={() => this.removeItem(index)} />
      );
    }
    if (column.key === 'edit') {
      return item.virtualPath === '/' ? null : (
        <IconButton iconProps={{ iconName: 'Edit' }} title="Edit" onClick={() => this._onShowPanel(item, index)} />
      );
    }
    if (column.key === 'type') {
      return item.virtualDirectory ? 'Directory' : 'Application';
    }
    return <span>{item[column.fieldName!]}</span>;
  };

  // tslint:disable-next-line:member-ordering
  private _columns: IColumn[] = [
    {
      key: 'virtualPath',
      name: 'Virtual Path',
      fieldName: 'virtualPath',
      minWidth: 210,
      maxWidth: 350,
      isRowHeader: true,
      data: 'string',
      isPadded: true,
      isResizable: true,
      ariaLabel: 'Operations for value',
    },
    {
      key: 'physicalPath',
      name: 'Physic Path',
      fieldName: 'physicalPath',
      minWidth: 210,
      maxWidth: 350,
      isRowHeader: true,
      data: 'string',
      isPadded: true,
      isResizable: true,
      ariaLabel: 'Operations for name',
    },

    {
      key: 'type',
      name: 'Type',
      fieldName: 'virtualDirectory',
      minWidth: 210,
      maxWidth: 350,
      isRowHeader: true,
      data: 'string',
      isPadded: true,
      isResizable: true,
      ariaLabel: 'Operations for value',
      onRender: this.onRenderItemColumn,
    },
    {
      key: 'delete',
      name: '',
      minWidth: 16,
      maxWidth: 16,
      isResizable: true,
      isCollapsable: false,
      onRender: this.onRenderItemColumn,
      ariaLabel: 'Operations for value',
    },
    {
      key: 'edit',
      name: '',
      minWidth: 16,
      maxWidth: 16,
      isResizable: true,
      isCollapsable: false,
      onRender: this.onRenderItemColumn,
      ariaLabel: 'Operations for value',
    },
  ];
}
