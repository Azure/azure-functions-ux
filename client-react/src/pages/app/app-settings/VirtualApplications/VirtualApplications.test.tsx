import { shallow } from 'enzyme';
import * as React from 'react';
import VirtualApplications from './VirtualApplications';
const VirtualApplicationsAny = VirtualApplications as any;
describe('Virtual Applications Table', () => {
  describe('Shallow Testing', () => {
    let container;

    beforeEach(() => {
      container = shallow(<VirtualApplicationsAny values={{ virtualApplications: [] }} />);
    });

    it('renders without crashing', () => {
      expect(container.length).toBe(1);
    });

    it('render matches snapshot', () => {
      expect(container).toMatchSnapshot('DefaultSnapshot');
    });
  });
});
