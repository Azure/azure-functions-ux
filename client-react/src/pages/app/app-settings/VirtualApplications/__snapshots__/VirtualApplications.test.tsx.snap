// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Virtual Applications Table Shallow Testing render matches snapshot: DefaultSnapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <VirtualApplications
    values={
      Object {
        "virtualApplications": Array [],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <CustomizedActionButton
          iconProps={
            Object {
              "iconName": "Add",
            }
          }
          onClick={[Function]}
          styles={
            Object {
              "root": Object {
                "marginTop": "5px",
              },
            }
          }
        >
          New Directory/Application
        </CustomizedActionButton>,
        <StyledPanelBase
          closeButtonAriaLabel="Close"
          headerText="New App "
          isOpen={false}
          onDismiss={[Function]}
          onRenderFooterContent={[Function]}
          type={3}
        >
          <VirtualApplicationsAddEdit
            updateVirtualApplication={[Function]}
          />
        </StyledPanelBase>,
        <StyledWithViewportComponent
          ariaLabelForSelectAllCheckbox="Toggle selection for all items"
          ariaLabelForSelectionColumn="Toggle selection"
          columns={
            Array [
              Object {
                "ariaLabel": "Operations for value",
                "data": "string",
                "fieldName": "virtualPath",
                "isPadded": true,
                "isResizable": true,
                "isRowHeader": true,
                "key": "virtualPath",
                "maxWidth": 350,
                "minWidth": 210,
                "name": "Virtual Path",
              },
              Object {
                "ariaLabel": "Operations for name",
                "data": "string",
                "fieldName": "physicalPath",
                "isPadded": true,
                "isResizable": true,
                "isRowHeader": true,
                "key": "physicalPath",
                "maxWidth": 350,
                "minWidth": 210,
                "name": "Physic Path",
              },
              Object {
                "ariaLabel": "Operations for value",
                "data": "string",
                "fieldName": "virtualDirectory",
                "isPadded": true,
                "isResizable": true,
                "isRowHeader": true,
                "key": "type",
                "maxWidth": 350,
                "minWidth": 210,
                "name": "Type",
                "onRender": [Function],
              },
              Object {
                "ariaLabel": "Operations for value",
                "isCollapsable": false,
                "isResizable": true,
                "key": "delete",
                "maxWidth": 16,
                "minWidth": 16,
                "name": "",
                "onRender": [Function],
              },
              Object {
                "ariaLabel": "Operations for value",
                "isCollapsable": false,
                "isResizable": true,
                "key": "edit",
                "maxWidth": 16,
                "minWidth": 16,
                "name": "",
                "onRender": [Function],
              },
            ]
          }
          isHeaderVisible={true}
          items={Array []}
          layoutMode={1}
          selectionMode={0}
          selectionPreservedOnEmptyClick={true}
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "New Directory/Application",
          "iconProps": Object {
            "iconName": "Add",
          },
          "onClick": [Function],
          "styles": Object {
            "root": Object {
              "marginTop": "5px",
            },
          },
        },
        "ref": null,
        "rendered": "New Directory/Application",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <VirtualApplicationsAddEdit
            updateVirtualApplication={[Function]}
          />,
          "closeButtonAriaLabel": "Close",
          "headerText": "New App ",
          "isOpen": false,
          "onDismiss": [Function],
          "onRenderFooterContent": [Function],
          "type": 3,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "updateVirtualApplication": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "ariaLabelForSelectAllCheckbox": "Toggle selection for all items",
          "ariaLabelForSelectionColumn": "Toggle selection",
          "columns": Array [
            Object {
              "ariaLabel": "Operations for value",
              "data": "string",
              "fieldName": "virtualPath",
              "isPadded": true,
              "isResizable": true,
              "isRowHeader": true,
              "key": "virtualPath",
              "maxWidth": 350,
              "minWidth": 210,
              "name": "Virtual Path",
            },
            Object {
              "ariaLabel": "Operations for name",
              "data": "string",
              "fieldName": "physicalPath",
              "isPadded": true,
              "isResizable": true,
              "isRowHeader": true,
              "key": "physicalPath",
              "maxWidth": 350,
              "minWidth": 210,
              "name": "Physic Path",
            },
            Object {
              "ariaLabel": "Operations for value",
              "data": "string",
              "fieldName": "virtualDirectory",
              "isPadded": true,
              "isResizable": true,
              "isRowHeader": true,
              "key": "type",
              "maxWidth": 350,
              "minWidth": 210,
              "name": "Type",
              "onRender": [Function],
            },
            Object {
              "ariaLabel": "Operations for value",
              "isCollapsable": false,
              "isResizable": true,
              "key": "delete",
              "maxWidth": 16,
              "minWidth": 16,
              "name": "",
              "onRender": [Function],
            },
            Object {
              "ariaLabel": "Operations for value",
              "isCollapsable": false,
              "isResizable": true,
              "key": "edit",
              "maxWidth": 16,
              "minWidth": 16,
              "name": "",
              "onRender": [Function],
            },
          ],
          "isHeaderVisible": true,
          "items": Array [],
          "layoutMode": 1,
          "selectionMode": 0,
          "selectionPreservedOnEmptyClick": true,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <CustomizedActionButton
            iconProps={
              Object {
                "iconName": "Add",
              }
            }
            onClick={[Function]}
            styles={
              Object {
                "root": Object {
                  "marginTop": "5px",
                },
              }
            }
          >
            New Directory/Application
          </CustomizedActionButton>,
          <StyledPanelBase
            closeButtonAriaLabel="Close"
            headerText="New App "
            isOpen={false}
            onDismiss={[Function]}
            onRenderFooterContent={[Function]}
            type={3}
          >
            <VirtualApplicationsAddEdit
              updateVirtualApplication={[Function]}
            />
          </StyledPanelBase>,
          <StyledWithViewportComponent
            ariaLabelForSelectAllCheckbox="Toggle selection for all items"
            ariaLabelForSelectionColumn="Toggle selection"
            columns={
              Array [
                Object {
                  "ariaLabel": "Operations for value",
                  "data": "string",
                  "fieldName": "virtualPath",
                  "isPadded": true,
                  "isResizable": true,
                  "isRowHeader": true,
                  "key": "virtualPath",
                  "maxWidth": 350,
                  "minWidth": 210,
                  "name": "Virtual Path",
                },
                Object {
                  "ariaLabel": "Operations for name",
                  "data": "string",
                  "fieldName": "physicalPath",
                  "isPadded": true,
                  "isResizable": true,
                  "isRowHeader": true,
                  "key": "physicalPath",
                  "maxWidth": 350,
                  "minWidth": 210,
                  "name": "Physic Path",
                },
                Object {
                  "ariaLabel": "Operations for value",
                  "data": "string",
                  "fieldName": "virtualDirectory",
                  "isPadded": true,
                  "isResizable": true,
                  "isRowHeader": true,
                  "key": "type",
                  "maxWidth": 350,
                  "minWidth": 210,
                  "name": "Type",
                  "onRender": [Function],
                },
                Object {
                  "ariaLabel": "Operations for value",
                  "isCollapsable": false,
                  "isResizable": true,
                  "key": "delete",
                  "maxWidth": 16,
                  "minWidth": 16,
                  "name": "",
                  "onRender": [Function],
                },
                Object {
                  "ariaLabel": "Operations for value",
                  "isCollapsable": false,
                  "isResizable": true,
                  "key": "edit",
                  "maxWidth": 16,
                  "minWidth": 16,
                  "name": "",
                  "onRender": [Function],
                },
              ]
            }
            isHeaderVisible={true}
            items={Array []}
            layoutMode={1}
            selectionMode={0}
            selectionPreservedOnEmptyClick={true}
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "New Directory/Application",
            "iconProps": Object {
              "iconName": "Add",
            },
            "onClick": [Function],
            "styles": Object {
              "root": Object {
                "marginTop": "5px",
              },
            },
          },
          "ref": null,
          "rendered": "New Directory/Application",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <VirtualApplicationsAddEdit
              updateVirtualApplication={[Function]}
            />,
            "closeButtonAriaLabel": "Close",
            "headerText": "New App ",
            "isOpen": false,
            "onDismiss": [Function],
            "onRenderFooterContent": [Function],
            "type": 3,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "updateVirtualApplication": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "ariaLabelForSelectAllCheckbox": "Toggle selection for all items",
            "ariaLabelForSelectionColumn": "Toggle selection",
            "columns": Array [
              Object {
                "ariaLabel": "Operations for value",
                "data": "string",
                "fieldName": "virtualPath",
                "isPadded": true,
                "isResizable": true,
                "isRowHeader": true,
                "key": "virtualPath",
                "maxWidth": 350,
                "minWidth": 210,
                "name": "Virtual Path",
              },
              Object {
                "ariaLabel": "Operations for name",
                "data": "string",
                "fieldName": "physicalPath",
                "isPadded": true,
                "isResizable": true,
                "isRowHeader": true,
                "key": "physicalPath",
                "maxWidth": 350,
                "minWidth": 210,
                "name": "Physic Path",
              },
              Object {
                "ariaLabel": "Operations for value",
                "data": "string",
                "fieldName": "virtualDirectory",
                "isPadded": true,
                "isResizable": true,
                "isRowHeader": true,
                "key": "type",
                "maxWidth": 350,
                "minWidth": 210,
                "name": "Type",
                "onRender": [Function],
              },
              Object {
                "ariaLabel": "Operations for value",
                "isCollapsable": false,
                "isResizable": true,
                "key": "delete",
                "maxWidth": 16,
                "minWidth": 16,
                "name": "",
                "onRender": [Function],
              },
              Object {
                "ariaLabel": "Operations for value",
                "isCollapsable": false,
                "isResizable": true,
                "key": "edit",
                "maxWidth": 16,
                "minWidth": 16,
                "name": "",
                "onRender": [Function],
              },
            ],
            "isHeaderVisible": true,
            "items": Array [],
            "layoutMode": 1,
            "selectionMode": 0,
            "selectionPreservedOnEmptyClick": true,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
